### DISCUSSION TITLE
Problem with use_module(library(option)) and :-debug(topic)

### DISCUSSION THREAD

**Discussion by hnbeck** [LogtalkDotOrg/logtalk3#123] [Q&A]
Hi all, 

at the moment I'm porting SWI Prolog code to a Logtalk object. Backend is SWI Prolog 8.3.22
In the source file in Logtalk, the line
:- use_module(library(option)).
results in error message    
```
! Domain error: value meta_options(1,:,-) is not in domain meta_argument_specifier
!       in clause for predicate meta_options/3
!       in auxiliary clause generated for file c:/dev/workspace/ectrafficmodeller/logic/xmpp/xmpp_config.lgt
!     
```
It is only with library(option), other libraries work as always.

Independently from this, the line
 :- debug(xmpp).
results in 

```
!     Domain error: value debug/1 is not in domain directive
!       in directive debug/1
!       in file c:/dev/workspace/ectrafficmodeller/logic/xmpp/xmpp_util.lgt at or above line 8
```

For the first case, I've read somethin in the SWI Group that there may be a problem with use_module/1 in some cases, but I've not understand the discussion
For the second case, I've no idea.
Can someone help me?

(BTW: there is nearly no code, I've just at the step setting up the file structure, containg of a loader file,  a Power shell script for compliation and the source file itself. :) )

**Comment by pmoura** [comment]
Note that Logtalk doesn't support a `use_module/1` directive, only its own `use_module/2` directive (which takes as first argument a module identifier instead of a file path). For details, see:

https://logtalk.org/manuals/refman/directives/use_module_2.html

True, there's code in some of the Prolog adapter files, including the one for SWI-Prolog, that try to expand Prolog `use_module/1` directives into Logtalk `use_module/2` directives. But that code doesn't work in all scenarios as is mainly intended to help in the compilation of modules as objects. One of the main issues is that the `:` meta-argument indicator is **ambiguous**, hence the error you got. Logtalk is not based on a predicate prefixing mechanism as found on modules systems. Thus, the compiler is unable to parse any argument declared as `:` as it can mean a goal, a closure, or neither. One possible solution is to override the problematic module `meta_predicate/1` directive as explained at:

https://logtalk.org/manuals/userman/predicates.html#calling-prolog-module-meta-predicates

An alternative solution would to use Logtalk's own `options` library:

https://logtalk.org/manuals/libraries/options.html

As this library is fully portable, that would be the best solution.

For debugging support, the Logtalk `debug_messages` tool provides similar functionality and is also fully portable:

https://logtalk.org/manuals/devtools/debug_messages.html

You get an error in that `debug/1` directive due to Logtalk not supporting (by design) the problematic use of arbitrary goals as directives.

**Reply by hnbeck** [reply]
Hi Paulo, 

great. I've found it and it worked. 

Now this problem beeing solved there is a problem which should not be one: 

`:- use_module(library(ssl), [ssl_context/3, ssl_negotiate/5, ssl_set_options/3]).`

results in the error 

```
!     Domain error: value ssl_context(+,-,:) is not in domain meta_argument_specifier
!       in clause for predicate ssl_context/3
!       in auxiliary clause generated for file c:/dev/workspace/ectrafficmodeller/logic/xmpp/xmpp_util.lgt
```
If I understand the links you provided me right this should work. Or is this a case where to override meta_predicate/1?

the object source file contains  no code, only some use_module things. The full list: 

```
:- use_module(library(pcre)).
	:- use_module(library(pprint)).
	:- use_module(library(ssl), [ssl_context/3,
								ssl_negotiate/5,
								ssl_set_options/3]).
	:- use_module(library(option), [option/2]).
	:- use_module(library(sgml), [new_sgml_parser/2,
									set_sgml_parser/2,
									sgml_parse/2,
									free_sgml_parser/1]).
	:- use_module(library(xpath), [xpath_chk/3]).
	:- use_module(library(apply), [maplist/3,
									exclude/3]).
	:- use_module(library(strings), [indent_lines/3]).	
```
and of course in the loader the same.

Cheers

Hans

**Reply by pmoura** [reply]
> Hi Paulo,
> 
> great. I've found it and it worked.
> 
> Now this problem beeing solved there is a problem which should not be one:
> 
> `:- use_module(library(ssl), [ssl_context/3, ssl_negotiate/5, ssl_set_options/3]).`
> 
> results in the error
> 
> ```
> !     Domain error: value ssl_context(+,-,:) is not in domain meta_argument_specifier
> !       in clause for predicate ssl_context/3
> !       in auxiliary clause generated for file c:/dev/workspace/ectrafficmodeller/logic/xmpp/xmpp_util.lgt
> ```
> 
> If I understand the links you provided me right this should work. Or is this a case where to override meta_predicate/1?

Note that `library(ssl)` is an abstraction of a file path, **not** the module name as required. Therefore:

```logtalk
:- use_module(ssl, [ssl_context/3, ssl_negotiate/5, ssl_set_options/3]).
```

And yes, you need to override the `meta_predicate/1` directive for that predicate. Maybe:

```logtalk
:- meta_predicate(ssl:ssl_context(*, *, *)).
```

> the object source file contains no code, only some use_module things. The full list:
> 
> ```
> :- use_module(library(pcre)).
> 	:- use_module(library(pprint)).
> 	:- use_module(library(ssl), [ssl_context/3,
> 								ssl_negotiate/5,
> 								ssl_set_options/3]).
> 	:- use_module(library(option), [option/2]).
> 	:- use_module(library(sgml), [new_sgml_parser/2,
> 									set_sgml_parser/2,
> 									sgml_parse/2,
> 									free_sgml_parser/1]).
> 	:- use_module(library(xpath), [xpath_chk/3]).
> 	:- use_module(library(apply), [maplist/3,
> 									exclude/3]).
> 	:- use_module(library(strings), [indent_lines/3]).	
> ```
> 
> and of course in the loader the same.

Usually, the loader file loads the modules using Prolog's `use_module/2` directives that avoid adding the module predicates to `user` space. For example:

```logtalk
:- use_module(library(ssl), []).
```

In a nutshell, Logtalk separates *loading* from *using*. In the particular case of Prolog modules, a main reason for this separation is that the modules must be loaded prior to compiling calls to the module predicates (specifically, to module meta-predicates). Lack of Prolog standardization, unfortunately, makes this a bit verbose.

**Reply by pmoura** [reply]
P.S. Note that `maplist/3` and `exclude/3` are available from Logtalk's portably [`meta`](https://logtalk.org/manuals/libraries/meta.html) library. Also, `option/2` is just `list::memberchk/2`.

**Reply by hnbeck** [reply]
Hi Paulo, 

Thanks again, all is solved. To be honest, I still don't understand every detail, term expansions and meta predicats are a little bit blurry to me :). But because I'm working now on a much bigger application now (Key words : Metaverse ready logical traffic simulation +, AR/VR visualization over an external engine) it is now time to read the Logtalk handbook again and in more detail. It is fascinating how much work you do and have done for compatibility and development. Therefore I'm now doing my XMPP stuff in Logtalk, and my goal is to get much more value out of Logtalk and its libraries. So any hint like yours above is warmly welcomed :)

Cheers + thanks

Hans

**Reply by pmoura** [reply]
Good to know that all issues are sorted out. Let's do a Q&A Logtalk webinar in January on those topics.

**Reply by hnbeck** [reply]
That would be a good idea. Here a unstructured collection of points which would interest me from my knowledge today:

- difference (or better) consequence of this and self
- where goes the memory? -> what happens at instantiaton of objects, "persistence" of data (in the sense of assert)
- examples ("Patterns") for using categories and protocols
- help for the developers -> tools for the developers provided by Logtalk
- Logtalk way of building knowlegde heavy systems (->  "patterns" or best practice)
- Using the multitask system and event library for building responsible servers (-> "patterns")

You may know, I'm coding in Prolog now for 4 years. I was also Smalltalker for many years. Now my overall question is to use Logtalk at the best, because it is a great system. But I'm struggling a little bit to "harmonize" my Smalltalk knowledge with my Prolog knowledge :)
My usage of Logtalk is limited up to now using simple objects with no explicit instantiation, now I want more :)

Cheers

Hans

**Reply by pmoura** [reply]
Those are all interesting topics to discuss. I will lookup and suggest a date for the webinar in January. Meanwhile, some tips on some of those topics:

- on `this` and `self`, see (if not already) the [`self_vs_this `](https://github.com/LogtalkDotOrg/logtalk3/tree/master/examples/self_vs_this) programming example; also their [glossary](https://logtalk.org/manuals/glossary.html) entries
- "[object ] instantiation" is a relation between objects that doesn't necessarily imply any sort of side effects (both the object playing the role of the *class* and the object playing the role of the *instance* can and often are static)
- for a gentle introduction to *protocols* and *categories*, see the [`planets`](https://github.com/LogtalkDotOrg/logtalk3/tree/master/examples/planets) programming example, notably the comments in the [`planets.lgt`](https://github.com/LogtalkDotOrg/logtalk3/blob/master/examples/planets/planets.lgt) source file.
- if you're familiar with *design patterns*,  there's also a [`design_patterns`](https://github.com/LogtalkDotOrg/logtalk3/tree/master/examples/design_patterns) programming example covering the most common ones

Given your knowledge of Smalltalk, your feedback on the [Smalltalk nomenclature](https://logtalk.org/manuals/userman/nomenclature.html#smalltalk-nomenclature) section on the Handbook would be much appreciated.

**Reply by hnbeck** [reply]
Thx for the pointers, and I'm looking forward to a new year full of Logtalk :) 
So Happy New Year to all :)

Hans
